ghost/core/core/server/services/collections/BookshelfCollectionsRepository.js

```diff
@@ -201,46 +204,46 @@ module.exports = class BookshelfCollectionsRepository {
+                });
                 await this.#relationModel.query().delete().where('collection_id', collection.id).transacting(options.transaction);
-            } else {
-                const collectionPostsOptions = {
-                    transaction: options.transaction,
-                    columns: ['id', 'post_id']
-                };
-                const existingRelations = await this.#fetchCollectionPostIds(collection.id, collectionPostsOptions);
-
-                for (const existingRelation of existingRelations) {
-                    const found = collectionPostsRelations.find((thing) => {
-                        return thing.post_id === existingRelation.post_id;
-                    });
-                    if (found) {
-                        found.id = null;
-                    } else {
-                        collectionPostRelationsToDeleteIds.push(existingRelation.id);
-                    }
+                if (collectionPostsRelations.length > 0) {
+                    await this.#relationModel.query().insert(collectionPostsRelations).transacting(options.transaction);
                 }
-            }
+            } else {
+                const collectionPostsToDelete = collection.events.filter(event => event.type === 'CollectionPostRemoved').map((event) => {
+                    return event.data.post_id;
+                });
 
-            const missingCollectionPostsRelations = collectionPostsRelations.filter(thing => thing.id !== null);
+                const collectionPostsToInsert = collection.events.filter(event => event.type === 'CollectionPostAdded').map((event) => {
+                    return {
+                        id: (new ObjectID).toHexString(),
+                        sort_order: 0,
+                        collection_id: collection.id,
+                        post_id: event.data.post_id
+                    };
+                });
 
-            if (missingCollectionPostsRelations.length > 0) {
-                await this.#relationModel.query().insert(missingCollectionPostsRelations).transacting(options.transaction);
-            }
-            if (collectionPostRelationsToDeleteIds.length > 0) {
-                await this.#relationModel.query().delete().whereIn('id', collectionPostRelationsToDeleteIds).transacting(options.transaction);
+                if (collectionPostsToDelete.length > 0) {
+                    await this.#relationModel.query().delete().where('collection_id', collection.id).whereIn('post_id', collectionPostsToDelete).transacting(options.transaction);
+                }
+                if (collectionPostsToInsert.length > 0) {
+                    await this.#relationModel.query().insert(collectionPostsToInsert).transacting(options.transaction);
+                }
             }
+
+            options.transaction.executionPromise.then(() => {
+                for (const event of collection.events) {
+                    this.#DomainEvents.dispatch(event);
```

Comment:
Why do we need to dispatch these? Nobody's listening, right? 📞

Reply:
No but I think it is the correct thing to do, for when someone does want to listen!